Exploration Node 1 : Classification
================
## 1. load_digits

#### 1-(1) load_digits import

In [1]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#### 1-(2) 데이터 준비

In [2]:
digits=load_digits()

#### 1-(3) 데이터 이해하기
##### 루브릭 1: 3가지 데이터셋의 구성을 파악하고 데이터를 이해는 과정이 포함되어 있는가?(상세기준: feature와 label선정을 위한 데이터 분석과정을 전개함)

dir()로 확인한 결과 데이터셋은 DESCR, data, feature_names, frame, images, target, target_names로 구성되어 있으며
각각의 형태는 다음과 같다. 데이터의 전체 구성을 확인하기 위하여 dir() 외에도 내부의 값을 확인하거나 shape을 이용하여 형태를 확인하였다.



In [3]:
print(dir(digits))

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']


In [4]:
print(digits.data.shape)
print(digits.data[0])
print(digits.feature_names)
print(digits.target.shape)
print(digits.DESCR)

(1797, 64)
[ 0.  0.  5. 13.  9.  1.  0.  0.  0.  0. 13. 15. 10. 15.  5.  0.  0.  3.
 15.  2.  0. 11.  8.  0.  0.  4. 12.  0.  0.  8.  8.  0.  0.  5.  8.  0.
  0.  9.  8.  0.  0.  4. 11.  0.  1. 12.  7.  0.  0.  2. 14.  5. 10. 12.
  0.  0.  0.  0.  6. 13. 10.  0.  0.  0.]
['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7',

feature 선정 : data
DESCR을 통해 데이터 개요를 살펴보면 instance가 1797개이며, attribute가 64개로 표현되었다.
attribute information에는 8*8 이미지 픽셀이 0~16의 값을 지니고 있다고 되어 있다.
모델을 만드는데 입력할 feature는 이미지 픽셀이 데이터화 된 값이 들어가야 하는데,
data의 shape값을 확인해 보니 (1797, 64)임을 알 수 있었다. 이는 DESCR에서 표현한
instance, attribute의 개수와 동일하기 때문에 이 값이 실제로 모델을 형성하는데 사용할 feature
라고 결론내렸다.
 
label 선정 : target
instance가 1797개이기 때문에 label역시 필연적으로 1797개일 수 밖에 없다. 여기서 feature인
data를 제외하고 1797개의 값(shape)를 갖는 변수는 target이 유일하기 때문에 target이 label임을
유추할 수 있다.

이후 노드에서 제시한 아래의 4가지 과정을 실행하였다.  
1. feature data 지정하기  
2. label data 지정하기  
3. target names 출력해 보기  
4. 데이터 describe 해보기

In [5]:
digits_data=digits.data
digits_label= digits.target
print(digits.target_names)
print(digits.DESCR)

[0 1 2 3 4 5 6 7 8 9]
.. _digits_dataset:

Optical recognition of handwritten digits dataset
--------------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 1797
    :Number of Attributes: 64
    :Attribute Information: 8x8 image of integer pixels in the range 0..16.
    :Missing Attribute Values: None
    :Creator: E. Alpaydin (alpaydin '@' boun.edu.tr)
    :Date: July; 1998

This is a copy of the test set of the UCI ML hand-written digits datasets
https://archive.ics.uci.edu/ml/datasets/Optical+Recognition+of+Handwritten+Digits

The data set contains images of hand-written digits: 10 classes where
each class refers to a digit.

Preprocessing programs made available by NIST were used to extract
normalized bitmaps of handwritten digits from a preprinted form. From a
total of 43 people, 30 contributed to the training set and different 13
to the test set. 32x32 bitmaps are divided into nonoverlapping blocks of
4x4 and the number of on pixels a

#### 1-(4) train, test 데이터 분리

In [6]:
X_train, X_test, y_train, y_test= train_test_split(digits_data, digits_label, test_size=0.3, random_state=5)

#### 1-(5) 다양한 모델로 학습시켜보기-
##### 루브릭 2: 3가지 데이터셋에 대해 각각 5가지 모델을 이용하여 학습을 수행하고 결과값을 얻었는가?(상세기준: 모델학습 및 테스트가 정상적으로 수행되고 결과값을 얻었음)

In [7]:
# decision tree 사용해보기
from sklearn.tree import DecisionTreeClassifier

decision_tree= DecisionTreeClassifier(random_state=20)
decision_tree.fit(X_train, y_train)
y_pred= decision_tree.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.91      0.95        58
           1       0.79      0.88      0.84        52
           2       0.89      0.83      0.86        58
           3       0.91      0.90      0.91        59
           4       0.80      0.81      0.80        43
           5       0.85      0.88      0.86        64
           6       0.98      0.89      0.93        47
           7       0.96      0.83      0.89        59
           8       0.77      0.80      0.78        50
           9       0.78      0.94      0.85        50

    accuracy                           0.87       540
   macro avg       0.87      0.87      0.87       540
weighted avg       0.87      0.87      0.87       540



In [8]:
# random forest 사용해 보기
from sklearn.ensemble import RandomForestClassifier

random_forest= RandomForestClassifier(random_state=20)
random_forest.fit(X_train, y_train)
y_pred= random_forest.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.98      0.99        58
           1       0.93      0.98      0.95        52
           2       1.00      1.00      1.00        58
           3       1.00      0.93      0.96        59
           4       0.98      1.00      0.99        43
           5       0.95      0.97      0.96        64
           6       1.00      0.98      0.99        47
           7       0.98      0.98      0.98        59
           8       0.96      0.94      0.95        50
           9       0.92      0.96      0.94        50

    accuracy                           0.97       540
   macro avg       0.97      0.97      0.97       540
weighted avg       0.97      0.97      0.97       540



In [9]:
# SVM 사용해 보기
from sklearn import svm

svm_model=svm.SVC()
svm_model.fit(X_train, y_train)
y_pred= svm_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        58
           1       0.98      1.00      0.99        52
           2       1.00      1.00      1.00        58
           3       1.00      0.97      0.98        59
           4       1.00      1.00      1.00        43
           5       0.97      0.98      0.98        64
           6       1.00      1.00      1.00        47
           7       1.00      0.98      0.99        59
           8       0.96      0.96      0.96        50
           9       0.94      0.96      0.95        50

    accuracy                           0.99       540
   macro avg       0.99      0.99      0.99       540
weighted avg       0.99      0.99      0.99       540



In [10]:
# SGD classifier 사용해 보기
from sklearn.linear_model import SGDClassifier
sgd_model=SGDClassifier()
sgd_model.fit(X_train, y_train)
y_pred= sgd_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        58
           1       0.98      0.90      0.94        52
           2       0.97      0.98      0.97        58
           3       1.00      0.88      0.94        59
           4       0.96      1.00      0.98        43
           5       0.94      0.94      0.94        64
           6       1.00      0.96      0.98        47
           7       1.00      0.95      0.97        59
           8       0.87      0.92      0.89        50
           9       0.80      0.96      0.87        50

    accuracy                           0.95       540
   macro avg       0.95      0.95      0.95       540
weighted avg       0.95      0.95      0.95       540



In [11]:
# logistic regression 사용해 보기
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(max_iter=3000)
logistic_model.fit(X_train, y_train)
y_pred= logistic_model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        58
           1       0.91      0.96      0.93        52
           2       0.98      1.00      0.99        58
           3       0.98      0.93      0.96        59
           4       0.95      0.93      0.94        43
           5       0.98      0.95      0.97        64
           6       1.00      0.96      0.98        47
           7       0.98      0.97      0.97        59
           8       0.94      0.96      0.95        50
           9       0.91      0.98      0.94        50

    accuracy                           0.96       540
   macro avg       0.96      0.96      0.96       540
weighted avg       0.97      0.96      0.96       540



C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:444: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


#### 1-(6) 모델을 평가해 보기: 예측 결과는 어떻게 해석하는가? 모델 성능평가 지료는 무엇이 좋은가? sklearn.metric에서 제공하는 평가지표 중 적절한 것을 선택하라.
##### 루브릭 3: 3가지 데이터셋에 대해 모델의 평가 지표를 선택하고 그 이유를 근거있게 서술하였는가?

궁극적으로는 f1이 가장 좋은 지표로 사용되어야 한다고 본다. 각 숫자는 가치중립적이다.
9가 0보다 더 좋거나 한 것이 아니라 0부터 9까지 어떤 카테고리든 같은 가치를 가지기 때문에
precision이나 recall 중 하나를 지표로 삼는 것은 f1이나 accuracy를 지표를 삼는 것에 비해 
부적절하다. 그렇기 때문에 여러 요소들을 고려할 때 precision과 recall값이 모두 반영된 f1 score가
가장 적절한 지표이며 이를 기준으로 모델을 평가할 것이다.  
다만, 학습 및 테스트에 사용된 label이 불균형을 이루지 않는 것으로 보이기 때문에
accracy도 사용 가능할 것으로 보인다.(하지만 그렇다고 recall이나 precision보다
뛰어난 지표라 할 수 있는가?? 하는 의문이 있다.)  
또한 각 카테고리에 대한 f1 score를 가중평균한 weighted avg를 기준으로 가장 높은 값을 정하여야
테스트에 사용한 레이블의 각 카테고리별 수가 달라도(각 카테고리의 support 개수가 달라도)
이를 반영한 점수를 선정할 수 있기 때문에 f1 score-weighted avg를 기준으로 가장 점수가 높은
모델인 svm을 digits데이터셋에 적합한 모델이라 결론을 내릴 것이다.

## 2. load_wine

#### 2-(1) load_wine import

In [12]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#### 2-(2) 데이터 준비

In [13]:
wine=load_wine()

#### 2-(3) 데이터 이해하기
##### 루브릭 1: 3가지 데이터셋의 구성을 파악하고 데이터를 이해는 과정이 포함되어 있는가?(상세기준: feature와 label선정을 위한 데이터 분석과정을 전개함)

dir()로 확인한 결과 데이터셋은 DESCR, data, feature_names, frame, images, target, target_names로 구성되어 있으며
각각의 형태는 다음과 같다. 데이터의 전체 구성을 확인하기 위하여 dir() 외에도 내부의 값을 확인하거나 shape을 이용하여 형태를 확인하였다.

In [14]:
print(dir(wine))

['DESCR', 'data', 'feature_names', 'frame', 'target', 'target_names']


In [15]:
print(wine.data.shape)
print(wine.data[0])
print(wine.feature_names)
print(wine.target.shape)
print(wine.DESCR)

(178, 13)
[1.423e+01 1.710e+00 2.430e+00 1.560e+01 1.270e+02 2.800e+00 3.060e+00
 2.800e-01 2.290e+00 5.640e+00 1.040e+00 3.920e+00 1.065e+03]
['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
(178,)
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====


feature 선정 : data
digits와 유사한 논리로 feature을 선정하였다. DESCR을 통해 데이터를 파악하여 instance가 178개이며, 
attribute는 13개, class(카테고리)는 3개임을 알 수 있었다. feature로 사용할 데이터는 총 178개이며,
13개의 속성을 가지고 있어야 하는데, 이러한 조건에 부합하는 shape을 가진 데이터는 data뿐이었기 때문에 data를
feature로 선정했다.
 
label 선정 : target
class(카테고리)가 0, 1, 2의 값을 가지고, instance가 178개 이므로 label은 (178,1)의 형태를 가지며
동시에 0, 1, 2중 하나의 값을 취하여야 한다. 이러한 조건에 부합하는 데이터는 target뿐이므로 target을
feature로 선정했다.

이후 노드에서 제시한 아래의 4가지 과정을 실행하였다.  
1. feature data 지정하기  
2. label data 지정하기  
3. target names 출력해 보기  
4. 데이터 describe 해보기

In [16]:
wine_data= wine.data
wine_label= wine.target
print(wine.target_names)
print(wine.DESCR)

['class_0' 'class_1' 'class_2']
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:           

#### 2-(4) train, test 데이터 분리

In [17]:
X_train2, X_test2, y_train2, y_test2= train_test_split(wine_data, wine_label, test_size=0.3, random_state=5)

#### 2-(5) 다양한 모델로 학습시켜보기-
##### 루브릭 2: 3가지 데이터셋에 대해 각각 5가지 모델을 이용하여 학습을 수행하고 결과값을 얻었는가?(상세기준: 모델학습 및 테스트가 정상적으로 수행되고 결과값을 얻었음)

In [18]:
# decision tree 사용해보기
from sklearn.tree import DecisionTreeClassifier

decision_tree= DecisionTreeClassifier(random_state=20)
decision_tree.fit(X_train2, y_train2)
y_pred2= decision_tree.predict(X_test2)
print(classification_report(y_test2, y_pred2))

              precision    recall  f1-score   support

           0       0.95      0.91      0.93        23
           1       0.89      0.89      0.89        18
           2       0.93      1.00      0.96        13

    accuracy                           0.93        54
   macro avg       0.92      0.93      0.93        54
weighted avg       0.93      0.93      0.93        54



In [19]:
# random forest 사용해 보기

from sklearn.ensemble import RandomForestClassifier

random_forest= RandomForestClassifier(random_state=20)
random_forest.fit(X_train2, y_train2)
y_pred2= random_forest.predict(X_test2)
print(classification_report(y_test2, y_pred2))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        23
           1       1.00      0.94      0.97        18
           2       0.93      1.00      0.96        13

    accuracy                           0.98        54
   macro avg       0.98      0.98      0.98        54
weighted avg       0.98      0.98      0.98        54



In [20]:
# SVM 사용해 보기
from sklearn import svm

svm_model=svm.SVC()
svm_model.fit(X_train2, y_train2)
y_pred2= svm_model.predict(X_test2)
print(classification_report(y_test2, y_pred2))

              precision    recall  f1-score   support

           0       0.89      0.74      0.81        23
           1       0.60      0.83      0.70        18
           2       0.40      0.31      0.35        13

    accuracy                           0.67        54
   macro avg       0.63      0.63      0.62        54
weighted avg       0.68      0.67      0.66        54



In [21]:
# SGD classifier 사용해 보기
from sklearn.linear_model import SGDClassifier

sgd_model=SGDClassifier()
sgd_model.fit(X_train2, y_train2)
y_pred2= sgd_model.predict(X_test2)
print(classification_report(y_test2, y_pred2))

              precision    recall  f1-score   support

           0       1.00      0.57      0.72        23
           1       0.69      0.61      0.65        18
           2       0.36      0.69      0.47        13

    accuracy                           0.61        54
   macro avg       0.68      0.62      0.61        54
weighted avg       0.74      0.61      0.64        54



In [22]:
# logistic regression 사용해 보기
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(max_iter=3000)
logistic_model.fit(X_train2, y_train2)
y_pred2= logistic_model.predict(X_test2)
print(classification_report(y_test2, y_pred2))

              precision    recall  f1-score   support

           0       1.00      0.96      0.98        23
           1       0.89      0.94      0.92        18
           2       0.92      0.92      0.92        13

    accuracy                           0.94        54
   macro avg       0.94      0.94      0.94        54
weighted avg       0.95      0.94      0.94        54



#### 2-(6) 모델을 평가해 보기: 예측 결과는 어떻게 해석하는가? 모델 성능평가 지료는 무엇이 좋은가? sklearn.metric에서 제공하는 평가지표 중 적절한 것을 선택하라.
##### 루브릭 3: 3가지 데이터셋에 대해 모델의 평가 지표를 선택하고 그 이유를 근거있게 서술하였는가?

f1지표가 가장 모델 평가에 바람직하다. 와인의 class가 뜻하는 것이 무엇인지는 정확히 알지 못하지만
단순히 와인의 종류를 나타낸 것이라면 digits 데이터셋처럼 가치중립적일 것이고
같은 종의 와인 중 고급, 중급, 저급으로 나눈 것이라면 급에 따라 나름의 가치를 지닌다고 할 수는 있겠지만
어떤 class가 고급, 중급 저급인지 알 수 없기 때문에 precision이나 recall보다 
종합적인 성능을 대변하는 f1지표가 모델 평가에 가장 바람직하다고 할 수 있다.
또한 각 class의 레이블 개수(support)가 다르기 때문에 weighted avg된 f1지표를 사용하는 것이 좋다.
그 결과 가장 높은 점수를 받은 모델인 random forest모델이 wine데이터셋에 가장 적합하다 할  것이다.

## 3. load_breast_cancer

#### 3-(1) load_breast_cancer import

In [23]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

#### 3-(2) 데이터 준비

In [24]:
breast_cancer=load_breast_cancer()

#### 3-(3) 데이터 이해하기
##### 루브릭 1. 3가지 데이터셋의 구성을 파악하고 데이터를 이해는 과정이 포함되어 있는가?(상세기준: feature와 label선정을 위한 데이터 분석과정을 전개함)

dir()로 확인한 결과 데이터셋은 DESCR, data, feature_names, frame, images, target, target_names로 구성되어 있으며
각각의 형태는 다음과 같다. 데이터의 전체 구성을 확인하기 위하여 dir() 외에도 내부의 값을 확인하거나 shape을 이용하여 형태를 확인하였다.

In [25]:
print(dir(breast_cancer))

['DESCR', 'data', 'data_module', 'feature_names', 'filename', 'frame', 'target', 'target_names']


In [26]:
print(breast_cancer.data.shape)
print(breast_cancer.data[0])
print(breast_cancer.feature_names)
print(breast_cancer.target.shape)
print(breast_cancer.DESCR)

(569, 30)
[1.799e+01 1.038e+01 1.228e+02 1.001e+03 1.184e-01 2.776e-01 3.001e-01
 1.471e-01 2.419e-01 7.871e-02 1.095e+00 9.053e-01 8.589e+00 1.534e+02
 6.399e-03 4.904e-02 5.373e-02 1.587e-02 3.003e-02 6.193e-03 2.538e+01
 1.733e+01 1.846e+02 2.019e+03 1.622e-01 6.656e-01 7.119e-01 2.654e-01
 4.601e-01 1.189e-01]
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
(569,)
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**D

feature 선정 : data / label 선정 : target
digits, wine과 유사한 논리로 feature를 선정하였다. 이에 대해서는 이미 2회 논의했기 때문에
같은 근거를 다시 대지 않는다. 실제 분석을 하는 상황에서 digits과제와 wine과제가
같은 데이터원에게서 전달되고 같은 구성, 같은 이름의 키를 가지고 있다면 바로 data와 target부터 확인
하는 것이 당연할 것이며, DESCR을 통해 확인하고 검증하였을 때도 같은 결과가 나옴을 알 수
있다.

이후 노드에서 제시한 아래의 4가지 과정을 실행하였다.  
1. feature data 지정하기  
2. label data 지정하기  
3. target names 출력해 보기  
4. 데이터 describe 해보기

In [27]:
breast_cancer_data= breast_cancer.data
breast_cancer_label= breast_cancer.target
print(breast_cancer.target_names)
print(breast_cancer.DESCR)

['malignant' 'benign']
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        worst/largest values) of these features were computed for each image,
        resulting in 30 features.  For instanc

#### 3-(4) train, test 데이터 분리

In [28]:
X_train3, X_test3, y_train3, y_test3= train_test_split(breast_cancer_data, breast_cancer_label, test_size=0.3, random_state=5)

#### 3-(5) 다양한 모델로 학습시켜보기 
##### 루브릭 2: 3가지 데이터셋에 대해 각각 5가지 모델을 이용하여 학습을 수행하고 결과값을 얻었는가?(상세기준: 모델학습 및 테스트가 정상적으로 수행되고 결과값을 얻었음)

In [29]:
# decision tree 사용해보기
from sklearn.tree import DecisionTreeClassifier

decision_tree= DecisionTreeClassifier(random_state=20)
decision_tree.fit(X_train3, y_train3)
y_pred3= decision_tree.predict(X_test3)
print(classification_report(y_test3, y_pred3))

              precision    recall  f1-score   support

           0       0.90      0.90      0.90        61
           1       0.95      0.95      0.95       110

    accuracy                           0.93       171
   macro avg       0.92      0.92      0.92       171
weighted avg       0.93      0.93      0.93       171



In [30]:
# random forest 사용해 보기

from sklearn.ensemble import RandomForestClassifier

random_forest= RandomForestClassifier(random_state=20)
random_forest.fit(X_train3, y_train3)
y_pred3= random_forest.predict(X_test3)
print(classification_report(y_test3, y_pred3))

              precision    recall  f1-score   support

           0       0.97      0.95      0.96        61
           1       0.97      0.98      0.98       110

    accuracy                           0.97       171
   macro avg       0.97      0.97      0.97       171
weighted avg       0.97      0.97      0.97       171



In [31]:
# SVM 사용해 보기
from sklearn import svm

svm_model=svm.SVC()
svm_model.fit(X_train3, y_train3)
y_pred3= svm_model.predict(X_test3)
print(classification_report(y_test3, y_pred3))

              precision    recall  f1-score   support

           0       1.00      0.90      0.95        61
           1       0.95      1.00      0.97       110

    accuracy                           0.96       171
   macro avg       0.97      0.95      0.96       171
weighted avg       0.97      0.96      0.96       171



In [32]:
# SGD classifier 사용해 보기
from sklearn.linear_model import SGDClassifier

sgd_model=SGDClassifier()
sgd_model.fit(X_train3, y_train3)
y_pred3= sgd_model.predict(X_test3)
print(classification_report(y_test3, y_pred3))

              precision    recall  f1-score   support

           0       0.62      0.98      0.76        61
           1       0.99      0.67      0.80       110

    accuracy                           0.78       171
   macro avg       0.81      0.83      0.78       171
weighted avg       0.86      0.78      0.79       171



In [33]:
# logistic regression 사용해 보기
from sklearn.linear_model import LogisticRegression

logistic_model = LogisticRegression(max_iter=3000)
logistic_model.fit(X_train3, y_train3)
y_pred3= logistic_model.predict(X_test3)
print(classification_report(y_test3, y_pred3))

              precision    recall  f1-score   support

           0       0.98      0.97      0.98        61
           1       0.98      0.99      0.99       110

    accuracy                           0.98       171
   macro avg       0.98      0.98      0.98       171
weighted avg       0.98      0.98      0.98       171



#### 3-(6) 모델을 평가해 보기: 예측 결과는 어떻게 해석하는가? 모델 성능평가 지료는 무엇이 좋은가? sklearn.metric에서 제공하는 평가지표 중 적절한 것을 선택하라.
##### 루브릭 3: 3가지 데이터셋에 대해 모델의 평가 지표를 선택하고 그 이유를 근거있게 서술하였는가?

앞의 두 문제와 유방암 모델의 가장 큰 차이점은 카테고리가 가치중립적이지도 않고 카테고리 정보를 통해
그 가치를 파악할 수 있다는 것이다. 유방암은 생명을 위협하는 큰 질병이기 때문에
유방암 양성이 바람직할 일일 가능성은 매우 적기 때문에 이러한 부분을 고려하여 지표를 선정해야 한다.
또한, 음성 판정의 경우 이후 검진자는 의료행위에 주의를 기울이지 않을 가능성이 높고
양성의 경우에는 재검 및 건강에 각별한 주의를 기울이게 될 것이기 때문에
그러므로 recall을 지표로 사용하는 것이 적절하다

#### 회고
지금까지 세가지 종류의 scikit-learn 데이터를 살펴보았다.
개인적으로 가장 흥미를 가지고 있었던 부분은 svm과 logistic regression의 비교였다.
svm과 logistic Regression에 대해 찾아볼 때, 개념적으로 좌표평면상의 선을 긋는 방식으로 논리가
전개된다는 사실을 알았고, 혹시 거의 같은 방식의 분류인가 하는 생각이 들었기 때문이다.
그래서 f1 score로 비교를 시도해 보았지만 분석하는 도중에 이러한 비교가 비효율적인 접근임을
알아차렸다. 데이터 수도 적고, 단순히 지표가 같은 수라는 것만으로 판단을 내릴 수 있는
문제가 애초에 아니었기 때문이다.

한편, decision tree보다 random forest가 늘 성능이 앞섰다. 
svm과 logistic regression의 비교와는 다르게, random forest가 decision tree의
단점을 보완한 모델이었다는 것을 이미 학습하였기 때문에, 점수를 보고 배운 내용에
더욱 확신을 가질 수 있었다.

가장 난처했던 부분은 linear regression에서 반복적으로 warning이 발생한 것이다.
후에 itration설정이 부족했기 때문임을 확인하고 수정하였지만
영어 실력 부족으로 해석과 적용에 많은 시간을 보냈기 때문이다.
다시 한번 영어의 중요성을 실감할 수 있었다.

wine데이터에서는 '어디까지 편의적으로 해석할 수 있는가'에 대해 고민했다.
위의 모델 평가에는 나오지 않았지만, 만약 정보의 이용자가 와인 판매상이고 가장 높은 와인의
등급이 class 2라면 class 2의 recall지표가 가장 높게 나오는 모델을 인위적으로 선정한
다음 '중립적인 빅데이터 AI와인 판별기'라는 광고문구를 내세우며 광고효과와 자신의 잇속을
챙길 수도 있을 것이다. 거꾸로 와인 구매자 입장에서는 class 0의 recall지표가 가장 높게
나오는 모델을 적용하여 더 높은 등급의 와인을 사려고 시도할 수도 있을 것이다.
분명 우리는 '중립적'인 관점에서 모델의 점수를 측정했지만
'편향적'인 방식으로 연구 결과가 이용될 수도 있는 것이다.
생각해 볼 문제다.

마지막으로 안타까웠던 점은 시간이 부족했다는 것이다.
다른 사람들이 미리 프로젝트를 한 다음 이것저것 시도해 볼 때,
시간이 부족하여 노드를 보느라 그만큼 신경을 쓰지 못한 것이 아쉽다.
데이터 시각화를 시도하는 사람들도 있었고, 깔끔한 포멧으로 정리하여 프로젝트를 제출하는 
사람들도 있었지만 그러지 못하다 보니 그들이 부러웠다.
왜냐하면, 시각적으로 열등해 보이는 것들에 대해 대부분의 사람들은 탐구할 생각을 하지 않기 때문이다.
그런 의미에서 다음에는 보이는 것들에 더 신경써서 과제를 마무리 하도록 노력해 보겠다.

Reference: 아이펠(https://lms.aiffel.io/) 학습 노드-Exploration 1